### ISSUE TITLE
on-bt-download-complete does not work

### ISSUE THREAD

**Issue by hungaborhorvath** [tatsuhiro-t/aria2#475]
In relation to issue #382 https://github.com/tatsuhiro-t/aria2/issues/382, I wrote a short script which kills aria2c daemon, then restarts it. I would like to run this script every time a bittorrent download is completed. The script is in /usr/local/bin/aria2c_restart

However, if I put 

on-bt-download-complete=/usr/local/bin/aria2c_restart

into my config file, then aria2c daemon quits almost immediately after started. Further, if I use aria2rpc to add this option, for a particular torrent, then I get an error message about --on-bt-download-complete, and the download does not start. I am using self-compiled 1.19.2 version on raspbian jessie.

**Comment by tatsuhiro-t** [comment]
> ... into my config file, then aria2c daemon quits almost immediately after started.

Did you specify --enable-rpc option?
Did you get any error message from aria2?

> Further, if I use aria2rpc to add this option, for a particular torrent, then I get an error message about --on-bt-download-complete, and the download does not start.

Currently --on-bt-download-complete cannot be specified per download. It can only be specified in config file or command-line.

**Comment by hungaborhorvath** [comment]
Yes, I did specify enable-rpc=true in the config file. 

I just cleared the log and tried it again, here is the error message if I have on-bt-download-complete=/usr/local/bin/aria2_restart in my config file:  

2015-11-06 12:59:35.688707 [ERROR] [HttpListenCommand.cc:114] IPv4 RPC: failed to bind TCP port 6800
Exception: [SocketCore.cc:301] errorCode=1 Failed to bind a socket, cause: Address is already in use
2015-11-06 12:59:35.803255 [ERROR] [MultiUrlRequestInfo.cc:292] Exception caught
Exception: [DownloadEngineFactory.cc:215] errorCode=1 Failed to setup RPC server.

If I comment out the line about on-bt-download-complete, then aria2c starts properly in rpc mode. I can access it with both aria2rpc and with webui-aria2.

**Comment by tatsuhiro-t** [comment]
Log says that aria2 failed to bind socket to port 6800, which probably means aria2c or another process uses port 6800.  Could you check that another aria2 is not started already?

**Comment by hungaborhorvath** [comment]
I checked. No other aria2 processes run. In fact, I tried several times after each other: aria2 starts, and after a couple seconds quits with the error message above. If I comment out on-bt-download-complete line from config file, then it starts and remains online as it is supposed to.

**Comment by tatsuhiro-t** [comment]
I could not reproduce this even if I wrote on-bt-download-complete option in my config file along with enable-rpc=true.
Could you paste your config file here?  Please mask any confidential information it contains.

**Comment by hungaborhorvath** [comment]
I paste it here, however I am not sure how to turn Markdown off.... :-(

# configuration file for aria2c
#check-integrity=true
bt-detach-seed-only=true
# same as bt-force-encryption=true
bt-require-crypto=true
bt-min-crypto-level=arc4
# 
bt-max-peers=25
bt-max-open-files=2
max-concurrent-downloads=1
max-connection-per-server=1
disable-ipv6=true
#enable-dht=false # default is true
#enable-peer-exchange=false # deafult is true
dht-file-path=/media/USBHDD1/.aria2/dht.dat
dht-file-path6=/media/USBHDD1/.aria2/dht6.dat
dir=/media/USBHDD1/Torrent/
log-level=warn
log=/media/USBHDD1/.aria2/aria2.log
ca-certificate=/etc/ssl/certs/ca-certificates.crt
#check-certificate=false
split=2
continue=true
force-save=true
bt-save-metadata=true
rpc-save-upload-metadata=true
save-session=/media/USBHDD1/.aria2/aria2_session.txt
input-file=/media/USBHDD1/.aria2/aria2_session.txt
save-session-interval=60
auto-save-interval=0
daemon=true
enable-rpc=true
rpc-listen-all=true
rpc-user=.........
rpc-passwd=.........
listen-port=.........
dht-listen-port=.........
bt-enable-lpd=true
#max-overall-upload-limit=16K
#max-upload-limit=12K
#max-overall-download-limit=128K
#max-download-limit=64K
file-allocation=falloc
#bt-external-ip=
### this is needed ### 
#on-bt-download-complete=/usr/local/bin/aria2_restart
#on-download-complete=
seed-ratio=0.0
#seed-time=2880
bt-remove-unselected-file=true
bt-request-peer-speed-limit=4K # default=50K

**Comment by tatsuhiro-t** [comment]
Thank you for providing configuration file.
Unfortunately, I still could not reproduce this issue even with the above configuration file.
In my environment, aria2 started as RPC daemon successfully, with on-bt-download-complete option line in configuration file.

**Comment by hungaborhorvath** [comment]
I use it on a raspberry pi, os is raspbian jessie. I have compiled the version 1.19.2 from source as a simple user (the same user who runs it), and I should still have the log files if you tell me where to look. I could post them if that would help, just tell me which files.

**Comment by tatsuhiro-t** [comment]
I have no idea at this point.  But your log file was really strange.
> 2015-11-06 12:59:35.688707 [ERROR] [HttpListenCommand.cc:114] IPv4 RPC: failed to bind TCP port 6800
> Exception: [SocketCore.cc:301] errorCode=1 Failed to bind a socket, cause: Address is already in use
> 2015-11-06 12:59:35.803255 [ERROR] [MultiUrlRequestInfo.cc:292] Exception caught
Exception: [DownloadEngineFactory.cc:215] errorCode=1 Failed to setup RPC server.

Do you really get this error log only when adding "on-bt-download-complete" option?

**Comment by hungaborhorvath** [comment]
If I comment out the line about on-bt-download-complete, then I get exactly this error. I know, because I cleared the log file completely before testing. 

Would the make/install logs help? Where can I find those? 

Thank you for looking into this. 

As a sidenote: the reason I need this is that currently the control file is either written to disk every N seconds, or only when aria exits gracefully. And the problem is that the torrent files are considered downloaded files, even if they are complete, so having seeded files in the list would constantly use the disk, even if e.g. there is no network activity. Therefore currently control files for me are only written to disk when aria2 exits. And then there are a lot of powercuts in our neighbourhood, and then when aria2 restarts, then it complains about already downloaded files I need to manually add them with check-integrity, etc.....

I reported this in issue #382, that maybe control files could be written to disk only if changed, just as with the session file. Then I would not need this option with on-bt-download-complete.
